feat(sdk-metrics): added synchronous gauge to SDK

clintonb opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/master/CONTRIBUTING.md#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?

Split from #4528, relates to #4296

## Short description of the changes
Added SDK for _synchronous_ gauge.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A

## Checklist:

- [x] Followed the style guidelines of this project
- [ ] Unit tests have been added
- [ ] Documentation has been updated

clintonb commented:
FYI @pichlermarc

clintonb commented:
@pichlermarc given the SDK depends on the API, is the order wrong here? It seems we should update the API first.

pichlermarc commented:
@clintonb ah, this is the counter-intuitive situation I mentioned over at #4565. :slightly_smiling_face: 

Since `@opentelemetry/sdk-metrics` has a peer-dependency on `@opentelemetry/api` and that peer dependency allows for using API versions starting at 1.3.0, we can only use 1.3.0 types. Since `Gauge` will only be added in 1.9.0 we can't use it in this package (the current build failure simulates what would happen to a user when they have an older API - for instance 1.8.0 installed that does not have the Gauge type available). 

The workaround is to copy the Gauge type from the API package to the `@opentelemetry/sdk-metrics` package. Since it's then copied 1:1 it'll be type-compatible with the API's version. 

We're currently discussing ways to improve this situation with SDK 2.0 (many edge cases, difficult topic), but in the meantime this is the way forward to add the feature. :slightly_smiling_face:

pichlermarc commented on packages/sdk-metrics/src/types.ts:
```diff
@@ -23,3 +24,13 @@ export type CollectionOptions = CommonReaderOptions;
 
 export type ForceFlushOptions = CommonReaderOptions;
+
+// NOTE: This is here because of the strange dependency on api?
```
You can use a similar text that we used previously:

https://github.com/open-telemetry/opentelemetry-js/blob/fec6286127129ec0082e5dd080242501f15ea0af/packages/sdk-metrics/src/InstrumentDescriptor.ts#L45-L50

pichlermarc reviewed (commented):
One last comment, then I think this should be good to merge. :slightly_smiling_face: 
Thanks for working on this @clintonb :rocket:

clintonb commented:
@pichlermarc I've rebased on main, added the `@experimental` comment flag, and updated the changelog.

Please take another look, and thanks for shepherding this!

clintonb commented:
@pichlermarc this circular dependency continues to hamper my development. The sdk-metrics tests depend on api, so I cannot increase test coverage to meet the threshold for merging.

clintonb commented:
See https://github.com/open-telemetry/opentelemetry-js/pull/4528 for complete test coverage.
